File: ruby/test/androidAppTest.rb
require 'rubygems'
require 'test/unit'
require 'appium_lib'
require 'selenium-webdriver'
require './configs.rb'
extend Test::Unit::Assertions

class AndroidAppTest < Test::Unit::TestCase

  def setup
    desire_cap = DesiredCapabilities.new
    @driver = Appium::Driver.new(desire_cap.desired_caps_android_app)
    @driver.start_driver.manage.timeouts.implicit_wait = 120
    @wait = Selenium::WebDriver::Wait.new(:timeout => 120)
  end

  def teardown
    if !@driver.nil?
      begin
        @driver.driver_quit
      rescue Exception => e
        puts e.message
      ensure
        @driver.driver_quit
      end
    end
  end

  def test_android_app
    verify_search_questions_on_acura_support_community
    verify_search_iFixit_on_home_screen
  end

  def verify_search_questions_on_acura_support_community
    first_question = 'Acura MDX'
    second_question = 'Cruise Control'
    puts 'Test case 1: it should allow to search some questions on Acura Support Community'

=begin
    Steps:
     1. Click on "Car and Truck" Categories on Homepage
     2. Click on "Acura" Categories
    Expected:
     1. General Information is "Acura".
     2.Verify five devices below displays.
     + Acura Integra
     + Acura MDX
     + Acura RL
     + Acura TL
     + Acura TSX
=end

    @driver.find_element(:xpath, "//*[@resource-id='android:id/home']").click
    sleep 5
    @driver.find_element(:xpath, "//*[@text='Car and Truck']").click
    @driver.find_element(:xpath, "//*[@text='Acura']").click
    sleep 5

    acura_text = @driver.find_element(:xpath, "//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=1]").text
    acura_integra_text = @driver.find_element(:xpath, "//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=3]").text
    acura_mdx_text = @driver.find_element(:xpath, "//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=4]").text
    acura_rl_text = @driver.find_element(:xpath, "//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=5]").text
    acura_tl_text = @driver.find_element(:xpath, "//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=6]").text
    acura_tsx_text = @driver.find_element(:xpath, "//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=7]").text

    assert_equal(acura_text, 'Acura')
    assert_equal(acura_integra_text, 'Acura Integra')
    assert_equal(acura_mdx_text, 'Acura MDX')
    assert_equal(acura_rl_text, 'Acura RL')
    assert_equal(acura_tl_text, 'Acura TL')
    assert_equal(acura_tsx_text, 'Acura TSX')

=begin
    Steps:
     1. Click on "Acura" General Information
     2. Swipe from left to right, it will move to Answers tab
     3. Enter keyword 'Acura MDX'
     4. Click Search icon
    Expected: It should show at least 6 results.
=end

    get_attr_name1 = search_question(first_question)
    sleep 2

=begin
    Steps:
     1. Clear text on Search field
     2. Enter keyword 'Cruise Control'
     3. Click on Search icon
     4. Wait a few seconds to get returned result
     5. Close app
    Expected: It should show at least 1 result.
=end

    @driver.driver.navigate.back
    get_attr_name2 = search_question(second_question)
    sleep 2
    @driver.close_app
    get_attr_name1 = get_attr_name1.gsub(/[^0-9]/, '')
    assert_true(get_attr_name1.to_i >= 6, "The expected results are greater or equal to 6 results but the actual results are #{get_attr_name1}.")
    get_attr_name2 = get_attr_name2.gsub(/[^0-9]/, '')
    assert_true(get_attr_name2.to_i >= 1, "The expected results are greater or equal to 1 results but the actual results are #{get_attr_name2}.")
  end

  def verify_search_iFixit_on_home_screen
    puts 'Test case 2: it should allow to search iFixit on Home screen'
    
=begin
     Steps:
     1. Reopen iFixit app
     2. Click Search menu on the left menu bar
     3. Search keyword 'Macbook Pro 2015'
     4. Press Enter button on keyboard
    Expected: It should show at least 48 results.
     5. Select Devices item on the Guides/Devices dropdown
    Expected: It should show at least 5 results.
=end
   
    @driver.launch_app
    sleep 2
    @driver.find_element(:xpath, "//*[@text='Search']").click
    @driver.find_element(:xpath, "//*[@resource-id='com.dozuki.ifixit:id/abs__search_src_text']")
    .send_keys('Macbook Pro 2015')
    @driver.press_keycode(66);
    sleep 2
    first_result = @driver.find_element(:xpath, "//*[@resource-id='com.dozuki.ifixit:id/search_result_count']").text
    @driver.find_element(:xpath, "//*[@resource-id='android:id/text1' and @text='Guides']").click
    @driver.find_element(:xpath, "//*[@resource-id='android:id/text1' and @text='Devices']").click

    second_result = @driver.find_element(:xpath, "//*[@resource-id='com.dozuki.ifixit:id/search_result_count']").text

    first_result = first_result.gsub(/[^0-9]/, '')
    assert_true(first_result.to_i >= 47, "The expected results are greater or equal to 47 results but the actual results are #{first_result}.")
    second_result = second_result.gsub(/[^0-9]/, '')
    assert_true(second_result.to_i >= 5 , "The returned results are greater or equal to 5 results but the actual results are #{second_result}.")
  end

  def swipe(direction)
    # Get the size of screen.
    size = @driver.window_size
    startx = (size.width * 0.90).to_i
    # Find endx point which is at left side of screen.
    endx = (size.width * 0.09).to_i
    # Find vertical point where you wants to swipe. It is in middle of screen height.
    starty = (size.height / 2).to_i
    if direction == 'RightLeft'
      # Swipe from Right to Left.
      @driver.swipe({
          :start_x => startx,
          :start_y => starty,
          :end_x => endx,
          :end_y => starty,
          :duration => 200
      })
    end
    if direction == 'LeftRight'
      # Swipe from Left to Right.
      @driver.swipe({
          :start_x => endx,
          :start_y => starty,
          :end_x => startx,
          :end_y => starty,
          :duration => 200
      })
    end
    sleep 5
  end

  def search_question(question)
    @driver.find_element(:xpath, "//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=1]").click
    sleep 2
    has_image = @driver.find_element(:id, "com.dozuki.ifixit:id/topic_info_image").displayed?
    assert_equal(has_image, true)
    swipe('RightLeft')
    @wait.until {@driver.find_element(:id, 'answersSearch')}.displayed?
    @driver.find_element(:id, 'answersSearch').send_keys(question)
    @driver.find_element(:id, 'searchIcon').click
    @wait.until {@driver.find_element(:xpath, "//android.view.View[contains(@content-desc,'Acura questions')]")}.displayed?
    results = @driver.find_element(:xpath, "//android.view.View[contains(@content-desc,'questions') and @index=1]").attribute('name')
    sleep 2
    return results
  end

end

File: php/AndroidAppTest.php
<?php
require_once('PHPUnit/Extensions/AppiumTestCase.php');
include("configs.php");

class AndroidAppTest extends PHPUnit_Extensions_AppiumTestCase {

  public $firstQuestion = 'Acura MDX';
  public $secondQuestion = 'Cruise Control';

  public static $browsers = desired_caps_android_app;

  public function test_android_app() {
    $this->search_questions_on_Acura_Support_Community();
    $this->search_IFixit_on_home_screen();
  }

  public function search_questions_on_Acura_Support_Community() {
    print 'should allow to search some questions on Acura Support Community';
    $this->byId('android:id/home')->click();
    sleep(5);
    $this->byXpath("//*[@text='Car and Truck']")->click();
    sleep(3);
    $this->byXpath("//*[@text='Acura']")->click();
    sleep(2);

    $acuraText = $this->byXpath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=1]")->text();
    $acuraIntegraText = $this->byXpath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=3]")->text();
    $acuraMDXText = $this->byXpath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=4]")->text();
    $acuraRLText = $this->byXpath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=5]")->text();
    $acuraTLText = $this->byXpath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=6]")->text();
    $acuraTSXText = $this->byXpath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=7]")->text();
    $this->assertEquals('Acura', $acuraText);
    $this->assertEquals('Acura Integra', $acuraIntegraText);
    $this->assertEquals('Acura MDX', $acuraMDXText);
    $this->assertEquals('Acura RL', $acuraRLText);
    $this->assertEquals('Acura TL', $acuraTLText);
    $this->assertEquals('Acura TSX', $acuraTSXText);

    $get_attr_name1 = $this->search_question($this->firstQuestion);

    /*
    Steps: 
    1. Clear text on Search field
    2. Enter keyword 'Cruise Control' 
    3. Click on Search icon 
    4. Wait a few seconds to get returned result
    5. Close app

    Expected: It should show at least 1 result.
    */

    $this->back();
    sleep(2);
    $get_attr_name2 = $this->search_question($this->secondQuestion);

    $this->closeApp();

    $get_attr_name1 = preg_replace("/[^0-9]/", '', $get_attr_name1);
    $this->assertGreaterThanOrEqual((int)6, $get_attr_name1,
      'The expected results are greater or equal to 6 results.');

    $get_attr_name2 = preg_replace("/[^0-9]/", '', $get_attr_name2);
    $this->assertGreaterThanOrEqual((int)1, $get_attr_name2,
      'The expected results are greater or equal to 1 result.');
  }

  public function search_IFixit_on_home_screen() {
    print 'should allow to search iFixit on Home screen';
    $this->launchApp();
    sleep(5);
    $this->byXpath("//*[@resource-id='com.dozuki.ifixit:id/action_search']")->click();
    sleep(2);
    $this->byXpath("//*[@resource-id='com.dozuki.ifixit:id/abs__search_src_text']")->click();
    sleep(2);
    $this->keys('Macbook Pro 2015');
    sleep(2);
    $this->keyEvent(66);
    sleep(5);
    $first_result = $this->byId("com.dozuki.ifixit:id/search_result_count")->text();
    $this->byXpath("//*[@resource-id='android:id/text1' and @text='Guides']")->click();
    sleep(2);
    $this->byXpath("//*[@resource-id='android:id/text1' and @text='Devices']")->click();
    sleep(2);
    $second_result = $this->byId("com.dozuki.ifixit:id/search_result_count")->text();

    $first_result = preg_replace("/[^0-9]/", '', $first_result);
    $this->assertGreaterThanOrEqual((int)47, $first_result,
      'The expected results are greater or equal to 47 results.');

    $second_result = preg_replace("/[^0-9]/", '', $second_result);
    $this->assertGreaterThanOrEqual((int)5, $second_result,
      'The expected results are greater or equal to 5 result.');
  }

  public function search_question($question) {
    $this->byXpath("//*[@resource-id='com.dozuki.ifixit:id/topic_title' and @index=1]")->click();
    sleep(5);
    $element = $this->byXpath("//*[@resource-id='com.dozuki.ifixit:id/topic_info_image']");
    sleep(5);
    $this->swipe_on_element($element, 'RightLeft');
    sleep(5);
    $this->byXpath("//*[@resource-id='answersSearch']")->click();
    $this->keys($question);
    $this->byId('searchIcon')->click();
    sleep(5);
    return $this->byXpath("//android.view.View[contains(@content-desc,'questions') and @index=1]")->attribute('name');
  }

  public function swipe_on_element($element, $direction = 'RightLeft') {
    $size = $element->size();
    $startx = $size['width'] * 0.95;
    // Find endx point which is at left side of screen.
    $endx = $size['width'] * 0.10;
    // Find vertical point where you wants to swipe. It is in middle of screen height.
    $starty = ($size['height'] * 0.50) + 100;

    if ($direction == 'RightLeft') {
      $this->swipe($startx, $starty, $endx, $starty, 200);
    }

    if ($direction == 'LeftRight') {
      $this->swipe($endx, $starty, $startx, $starty, 200);
    }
    sleep(3);
  }
}